docs/function-table-stable-sort.md

```diff
@@ -0,0 +1,87 @@
+
+The [ways currently available](#Alternatives) to achieve a stable sort in Luau have signicant costs in time and memory. Also, though one doesn't always need a stable sort, the contexts where it is needed are many and various. If everyone has to implement a stable sort on their own, either from scratch or piggybacking on top of the unstable sort in the `table` library, their implementations will have less review and less thorough testing than a library version.
+
+I propose to instead provide a stable sorting algorithm in Luau's `table` library. It could initially be behind a feature flag, or offered in parallel to the existing unstable sort. If its performance can be made comparable to the existing algorithms, one may consider having it be the only sort method provided.
+
+
+## Design
+
+The best stable sorting algorithms I know of are Python's [timsort](https://en.wikipedia.org/wiki/Timsort) and [block merge sort](https://en.wikipedia.org/wiki/Block_sort). Both are O(n log n) worst-case algorithms. I favor block-sort because it has variants requiring only constant space overhead (and this overhead can be kept very small, at some performance cost). The C implementations I've been testing are based on code from the [Grail Sort project](https://github.com/HolyGrailSortProject/Rewritten-Grailsort).
+
+The proposed API would be:
+
+`table.stable_sort <V>: (t: {V}, lt: ((V, V) -> boolean)?, reverse: boolean?) -> ()`
+
+(Or alternatively, this could replace `table.sort` but be behind a feature flag.)
+
+The optional third argument inverts the direction of the comparison being used (whether that's the default `lua_lessthan`, or a user-supplied comparison function). The infrastructure for this is a natural part of the sorting implementation anyway, so it seems a shame not to expose it in the user-facing API. This can avoid the overhead of an extra Lua function call for each comparison, when all one wants to do is invert the order of a search. (It'd also be straightforward to expose such an optional third argument on the existing unstable `table.sort` function.)
+
+The block-sorting algorithm internally needs a table rotation algorithm, and we might as well also expose that as:
+
+`table.rotate <V>: (t: {V}, leftwards: number) -> {V}`
```

Comment:
Let's exclude this method from the proposal until there is motivation for including it that's maybe a bit more than 'might as well'.
Because library implementation is not going to call the Luau method and will rotate directly.